networkx/utils/decorators.py

```diff
@@ -454,22 +365,539 @@ def random_array(dims, random_state=1):
+        func._code = code
+        return func
+
+    def assemble(self, f):
+        """Collects the requisite data to compile the decorated version of f.
+
+        Note, this is recursive, and all argmap-decorators will be flattened
+        into a single function call
+
+        Parameters
+        ----------
+        f : callable
+            The function to be assembled
+
+        Returns
+        -------
+        sig : argmap.Signature
+            The function signature.
+        wrapped_name : str
+            The mangled name used to represent the wrapped function in the code
+            being assembled.
+        functions : dict
+            A dictionary mapping id(g) -> (mangled_name(g), g) for functions g
+            referred to in the code being assembled
+        mapblock : list of lists and/or strings
+            Code that implements mapping of parameters including try blocks
+        finallys : list of lists and/or strings
+            Code that implements the closing-functions in finally blocks
+        mutable_args : bool
+            True if the compiled function will attempt to modify positional
+            arguments via their indices.
+        """
+
+        # first, we check if f is already argmapped -- if that's the case,
+        # build up the function recursively.
+        # > mapblock is generally a list of function calls of the sort
+        #     arg = func(arg)
+        # in addition to some try-blocks.
+        # > finallys is a recursive list of finally blocks of the sort
+        #         finally:
+        #             close_func_1()
+        #     finally:
+        #         close_func_2()
+        # > functions is a dict of functions used in the scope of our decorated
+        # function.  It will be used to construct globals used in compilation.
+        # We make functions[id(f)] = name_of_f, f to ensure that a given
+        # function is stored and named exactly once.
+        if hasattr(f, "__argmap__") and f.__self__ is f:
+            (
+                sig,
+                wrapped_name,
+                functions,
+                mapblock,
+                finallys,
+                mutable_args,
+            ) = f.__argmap__.assemble(f.__wrapped__)
+            functions = dict(functions)  # shallow-copy just in case
+        else:
+            sig = self.signature(f)
+            wrapped_name = self._name(f)
+            mapblock, finallys = [], []
+            functions = {id(f): (wrapped_name, f)}
+            mutable_args = False
+
+        if id(self._func) in functions:
+            fname, _ = functions[id(self._func)]
+        else:
+            fname, _ = functions[id(self._func)] = self._name(self._func), self._func
+
+        # this is a bit complicated -- we can call functions with a variety of
+        # arguments, so long as their input and output are tuples with the same
+        # structure
+        applied = set()
+
+        def get_name(arg, first=True):
```

Comment:
reference the `is_partition` decorator which forced the existence of this feature